File: mobile/src/main/java/info/papdt/express/helper/ui/fragment/home/BaseFragment.java
package info.papdt.express.helper.ui.fragment.home;

import android.app.Activity;
import android.content.Context;
import android.os.AsyncTask;
import android.os.Bundle;
import android.os.Handler;
import android.os.Message;
import android.support.annotation.Nullable;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.View;
import android.widget.LinearLayout;

import com.h6ah4i.android.widget.advrecyclerview.swipeable.RecyclerViewSwipeManager;
import com.h6ah4i.android.widget.advrecyclerview.touchguard.RecyclerViewTouchActionGuardManager;

import com.scwang.smartrefresh.layout.api.RefreshLayout;
import com.scwang.smartrefresh.layout.listener.OnRefreshListener;
import info.papdt.express.helper.R;
import info.papdt.express.helper.dao.PackageDatabase;
import info.papdt.express.helper.ui.MainActivity;
import info.papdt.express.helper.ui.adapter.HomePackageListAdapter;
import info.papdt.express.helper.ui.callback.OnDataRemovedCallback;
import info.papdt.express.helper.ui.common.AbsFragment;
import info.papdt.express.helper.view.AnimatedRecyclerView;
import info.papdt.express.helper.view.DeliveryHeader;

public abstract class BaseFragment extends AbsFragment implements OnRefreshListener {

	private RefreshLayout mRefreshLayout;
	private AnimatedRecyclerView mRecyclerView;
	private LinearLayout mEmptyView;

	private RecyclerView.Adapter mAdapter;
	private RecyclerViewSwipeManager mSwipeManager;

	private PackageDatabase mDatabase;
	private static Context sInstance = null;
	private final static int FLAG_REFRESH_LIST = 0, FLAG_UPDATE_ADAPTER_ONLY = 1;
	private boolean hasPlayedAnimation = false;

	int eggCount = 0, bigEggCount = 0;

	public BaseFragment(PackageDatabase database) {
		this.mDatabase = database;
	}

	public BaseFragment(){
		super();
	}

	// official method to get Activity Context
	@Override
	public void onAttach(Activity activity) {
		super.onAttach(activity);
		sInstance = activity;
	}

	// restore database to reconstruct express info
	@Override
	public void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		this.mDatabase = PackageDatabase.Companion.getInstance(sInstance);
	}

	@Override
	protected int getLayoutResId() {
		return R.layout.fragment_home;
	}

	@Override
	protected void doCreateView(View rootView) {
		mRefreshLayout = $(R.id.refresh_layout);
		mRecyclerView = $(R.id.recycler_view);
		mEmptyView = $(R.id.empty_view);

		// Set up mRecyclerView
		mSwipeManager = new RecyclerViewSwipeManager();
		mRecyclerView.setHasFixedSize(false);
		mRecyclerView.setLayoutManager(
				new LinearLayoutManager(getContext(), LinearLayoutManager.VERTICAL, false)
		);

		// Set up mRefreshLayout
		mRefreshLayout.setOnRefreshListener(this);
		mRefreshLayout.setRefreshHeader(new DeliveryHeader(rootView.getContext()));

		setUpAdapter();
		mEmptyView.setVisibility(mAdapter != null && mAdapter.getItemCount() > 0 ? View.GONE : View.VISIBLE);
		mEmptyView.setOnClickListener(new View.OnClickListener() {
			@Override
			public void onClick(View view) {
				if (eggCount++ > 7) {
					eggCount = 0;
					bigEggCount++;
					View eggView = $(R.id.sun);
					if (eggView == null) return;
					eggView.setRotation(0f);
					eggView.animate().rotation(360f * bigEggCount).setDuration(1000).start();
					if (bigEggCount > 3) bigEggCount = 0;
				}
			}
		});
	}

	protected abstract void setUpAdapter();
	public abstract int getFragmentId();

	@Override
	public void onRefresh(RefreshLayout view) {
		mHandler.sendEmptyMessage(FLAG_REFRESH_LIST);
	}

	public void notifyDataSetChanged() {
		mHandler.sendEmptyMessage(FLAG_UPDATE_ADAPTER_ONLY);
	}

	public void scrollToTop() {
		if (mAdapter != null && mAdapter.getItemCount() > 0) {
			mRecyclerView.smoothScrollToPosition(0);
		}
	}

	protected PackageDatabase getDatabase() {
		return mDatabase;
	}

	protected void playListAnimation() {
		if (!hasPlayedAnimation) {
			hasPlayedAnimation = true;
			mRecyclerView.scheduleLayoutAnimation();
		}
	}

	protected void setAdapter(RecyclerView.Adapter adapter) {
		this.mAdapter = adapter;
		mRecyclerView.setAdapter(mSwipeManager.createWrappedAdapter(mAdapter));
		mSwipeManager.attachRecyclerView(mRecyclerView);
		new RecyclerViewTouchActionGuardManager().attachRecyclerView(mRecyclerView);
		mEmptyView.setVisibility(mAdapter != null && mAdapter.getItemCount() > 0 ? View.GONE : View.VISIBLE);

		/** Set undo operation */
		if (adapter instanceof HomePackageListAdapter) {
			((HomePackageListAdapter) adapter).setOnDataRemovedCallback(new OnDataRemovedCallback() {
				@Override
				public void onDataRemoved(int pos, @Nullable String title) {
					Message msg = new Message();
					msg.what = MainActivity.MSG_NOTIFY_ITEM_REMOVE;
					msg.arg1 = getFragmentId();
					Bundle data = new Bundle();
					data.putString("title", title);
					msg.setData(data);

					getMainActivity().mHandler.sendMessage(msg);

					getMainActivity().notifyDataChanged(getFragmentId());
				}
			});
		}
	}

	protected MainActivity getMainActivity() {
		return (MainActivity) getActivity();
	}

	public void onUndoActionClicked() {
		int position = mDatabase.undoLastRemoval();
		if (position >= 0 && mAdapter != null) {
			mAdapter.notifyDataSetChanged();
			getMainActivity().notifyDataChanged(getFragmentId());
		}
	}

	private Handler mHandler = new Handler() {
		@Override
		public void handleMessage(Message msg) {
			switch (msg.what) {
				case FLAG_REFRESH_LIST:
					if (mRefreshLayout != null && !mRefreshLayout.isRefreshing()) {
						mRefreshLayout.autoRefresh();
					}
					new RefreshTask().execute();
					break;
				case FLAG_UPDATE_ADAPTER_ONLY:
					if (mAdapter != null) {
						mAdapter.notifyDataSetChanged();
						playListAnimation();
						mEmptyView.setVisibility(mAdapter != null && mAdapter.getItemCount() > 0 ? View.GONE : View.VISIBLE);
					}
					break;
			}
		}
	};

	public class RefreshTask extends AsyncTask<Void, Void, Void> {

		@Override
		protected Void doInBackground(Void... voids) {
			mDatabase.pullDataFromNetwork(false);
			return null;
		}

		@Override
		protected void onPostExecute(Void msg) {
			hasPlayedAnimation = false;
			mRefreshLayout.finishRefresh();
			mHandler.sendEmptyMessage(FLAG_UPDATE_ADAPTER_ONLY);
		}

	}

}

File: mobile/src/main/kotlin/info/papdt/express/helper/ui/fragment/home/BaseFragment.kt
package info.papdt.express.helper.ui.fragment.home

import android.annotation.SuppressLint
import android.content.Context
import android.os.Bundle
import android.os.Handler
import android.os.Message
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.view.View
import android.widget.LinearLayout
import android.widget.TextView
import com.scwang.smartrefresh.header.DeliveryHeader

import com.scwang.smartrefresh.layout.SmartRefreshLayout

import com.scwang.smartrefresh.layout.api.RefreshLayout
import com.scwang.smartrefresh.layout.constant.RefreshState
import com.scwang.smartrefresh.layout.listener.OnRefreshListener
import info.papdt.express.helper.R
import info.papdt.express.helper.dao.PackageDatabase
import info.papdt.express.helper.ui.MainActivity
import info.papdt.express.helper.ui.common.AbsFragment
import info.papdt.express.helper.view.AnimatedRecyclerView
import io.reactivex.Observable
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers

abstract class BaseFragment : AbsFragment, OnRefreshListener {

	lateinit var mRefreshLayout: SmartRefreshLayout
	private lateinit var mRecyclerView: AnimatedRecyclerView
	private lateinit var mEmptyView: LinearLayout
	lateinit var mSwipeHeader: DeliveryHeader
	protected lateinit var mEmptyTips: TextView

	private var mAdapter: RecyclerView.Adapter<*>? = null

	protected var database: PackageDatabase? = null
		private set
	private var hasPlayedAnimation = false

	private var eggCount = 0
	private var bigEggCount = 0

	var isDemoRefresh = false

	constructor(database: PackageDatabase) {
		this.database = database
	}

	constructor() : super()

	// official method to get Activity Context
	override fun onAttach(context: Context?) {
		super.onAttach(context)
		sInstance = context
	}

	// restore database to reconstruct express info
	override fun onCreate(savedInstanceState: Bundle?) {
		super.onCreate(savedInstanceState)
		this.database = PackageDatabase.getInstance(sInstance!!)
	}

	override fun getLayoutResId(): Int {
		return R.layout.fragment_home
	}

	override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
		mRefreshLayout = view.findViewById(R.id.refresh_layout)
		mRecyclerView = view.findViewById(R.id.recycler_view)
		mEmptyView = view.findViewById(R.id.empty_view)
		mSwipeHeader = view.findViewById(R.id.refresh_header)
		mEmptyTips = view.findViewById(R.id.frame_empty_tip)

		// Set up mRecyclerView
		mRecyclerView.setHasFixedSize(false)
		mRecyclerView.layoutManager = LinearLayoutManager(context, LinearLayoutManager.VERTICAL, false)

		// Set up mRefreshLayout
		mRefreshLayout.setOnRefreshListener(this)
		mRefreshLayout.setRefreshHeader(DeliveryHeader(view.context))

		setUpAdapter()
		mEmptyView.visibility = if (mAdapter != null && mAdapter!!.itemCount > 0) View.GONE else View.VISIBLE
		mEmptyView.setOnClickListener(View.OnClickListener {
			if (eggCount++ > 7) {
				eggCount = 0
				bigEggCount++
				val eggView = `$`<View>(R.id.sun) ?: return@OnClickListener
				eggView.rotation = 0f
				eggView.animate().rotation(360f * bigEggCount).setDuration(1000).start()
				if (bigEggCount > 3) bigEggCount = 0
			}
		})
	}

	protected abstract fun setUpAdapter()
	abstract val fragmentId: Int

	override fun onRefresh(view: RefreshLayout) {
		if (isDemoRefresh) {
			isDemoRefresh = false
			return
		}
		mHandler.sendEmptyMessage(FLAG_REFRESH_LIST)
	}

	fun notifyDataSetChanged() {
		mHandler.sendEmptyMessage(FLAG_UPDATE_ADAPTER_ONLY)
	}

	fun scrollToTop() {
		if (mAdapter != null && mAdapter!!.itemCount > 0) {
			mRecyclerView.smoothScrollToPosition(0)
		}
	}

	protected fun playListAnimation() {
		if (!hasPlayedAnimation) {
			hasPlayedAnimation = true
			mRecyclerView.scheduleLayoutAnimation()
		}
	}

	protected fun setAdapter(adapter: RecyclerView.Adapter<*>) {
		this.mAdapter = adapter
		mRecyclerView.adapter = mAdapter
		mEmptyView.visibility = if (mAdapter != null && mAdapter!!.itemCount > 0) View.GONE else View.VISIBLE
	}

	protected val mainActivity: MainActivity
		get() = activity as MainActivity

	fun onUndoActionClicked() {
		val position = database!!.undoLastRemoval()
		if (position >= 0 && mAdapter != null) {
			mAdapter!!.notifyDataSetChanged()
			mainActivity.notifyDataChanged(fragmentId)
		}
	}

	private val mHandler = @SuppressLint("HandlerLeak")
	object : Handler() {
		override fun handleMessage(msg: Message) {
			when (msg.what) {
				FLAG_REFRESH_LIST -> {
					if (mRefreshLayout.state != RefreshState.Refreshing) {
						mRefreshLayout.autoRefresh()
					}
					Observable.just(false).map(database!!::pullDataFromNetwork)
							.subscribeOn(Schedulers.io())
							.observeOn(AndroidSchedulers.mainThread())
							.subscribe {
								hasPlayedAnimation = false
								mRefreshLayout.finishRefresh()
								sendEmptyMessage(FLAG_UPDATE_ADAPTER_ONLY)
							}
				}
				FLAG_UPDATE_ADAPTER_ONLY -> if (mAdapter != null) {
					mAdapter!!.notifyDataSetChanged()
					playListAnimation()
					mEmptyView.visibility = if (mAdapter != null && mAdapter!!.itemCount > 0) View.GONE else View.VISIBLE
				}
			}
		}
	}

	companion object {

		@SuppressLint("StaticFieldLeak")
		private var sInstance: Context? = null
		private const val FLAG_REFRESH_LIST = 0
		private const val FLAG_UPDATE_ADAPTER_ONLY = 1

	}

}
